## Setup

Run the code below to load the scrambled message:


In [1]:
import pandas as pd
import numpy as np

message = pd.read_csv("https://www.dropbox.com/s/lgpn3vmksk3ssdo/scrambled_message.txt?dl=1")['Word']

In this activity, a "word" refers to any set of characters with no white space, even though they are not truly an English word.  That is, even though many of elements of the scrambled message vector are nonsense, and some have punctuation, you can consider each element to be a "word".

Beware!  The object named `message` is a **pandas Series** of strings.  If you want to use functions that expect a string, rather than a series, remember `.apply()` and `lambda` functions.




## Warm-up exercises

1. How many characters are in the scrambled message?
2. How many of these characters are white space?
3. How many words are in the scrambled message?
4. Show all the punctuation marks in the scrambled message.
5. Print out, in all capitals, the longest word in the scrambled message.
6. Print out every piece of a word that starts with the letter "m" and ends with the letter "z" in the scrambled message.

In [2]:
message

0                    Koila!
1                     In   
2                     kiew,
3                         a
4                 humble   
               ...         
122                     you
123                 mabugh.
124              ughhh?call
125        meugh.ughhhh!   
126                      K.
Name: Word, Length: 127, dtype: object

In [3]:
long_string = ' '.join(message)
long_string
print(len(long_string))

2670


'Koila!                In                   kiew,                    a            humble    kaudevillianugh?aoghajdbn          kezeran,                 casz             kicariouslb                as                    bozh            kiczim               ughhh!and              killain                bb                  zhe            kicissizudes             ughhh!of                faze.          Thisughhhh!           kisage,            noughhhh!          ughhh?mereugh?            kenaar            ughhh?ofugh?        ughhh?kanizy,          ughhh?is                    a              keszige                      of               zhe                     kox      ughhh!populi                  now                 kacanz,     kanished.ugh?                However,                 zhis             kalorous        kisizazion                      of                 a               bygone                kexazion            szands               kivified,         ughhh!and                 hasugh. 

1. There are 2670 characters in the scrambled message

In [4]:
white_space = long_string.count(" ")
white_space

1778

There are 1778 white spaces in the message

In [7]:
import re
long_string1 = re.split(r"\s", long_string)
long_string1

['',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 'Koila!',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 'In',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 'kiew,',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 'a',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 'humble',
 '',
 '',
 '',
 'kaudevillianugh?aoghajdbn',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 'kezeran,',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 'casz',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 'kicariouslb',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 'as',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 'bozh',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 'kiczim',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',
 '',


In [ ]:
long_string_word = list(filter(None, long_string1))
long_string_word
print(len(long_string_word))

127


3. There are 127 words in the scrambles message

4. Show all the punctuation marks in the scrambled message:

In [ ]:
import string

for i in long_string:
    if i in string.punctuation:
        print("Punctuation: " + i)

5. Print out, in all capitals, the longest word in the scrambled message.

In [ ]:
for i in long_string_word:
  length = len(i)
  if len(i) == 25:
    print(i.upper())

KAUDEVILLIANUGH?AOGHAJDBN


6. Print out every piece of a word that starts with the letter "m" and ends with the letter "z" in the scrambled message.

In [ ]:
matches = re.findall(r'\bm\w*z\b', long_string)
matches

['maaz']



## Decode a message

Complete the following steps to decode the message.  

1. Remove any spaces before or after each word.
2. Any time you see the word "ugh", with any number of h's, followed by a punctuation mark, delete this.
3. No word should be longer than 16 characters. Drop all extra characters beyond 13 off the end of each word.
4. Replace all instances of exactly 2 a's with exactly 2 e's.
5. Replace all z's with t's.
6. Every word that ends in b, change that to a y.  *Hint: look out for punctuation!*
7. Every word that starts with k, change that to a v.  *Hint: look out for capitalization!*
8. Use `.join()` to recombine all your words into a message.
9. Find the movie this quote is from.

1. Remove any spaces before or after each word.

In [8]:
list_message = message.tolist()

list_message = list(map(lambda x: x.replace(" ", ""), list_message))

list_message

['Koila!',
 'In',
 'kiew,',
 'a',
 'humble',
 'kaudevillianugh?aoghajdbn',
 'kezeran,',
 'casz',
 'kicariouslb',
 'as',
 'bozh',
 'kiczim',
 'ughhh!and',
 'killain',
 'bb',
 'zhe',
 'kicissizudes',
 'ughhh!of',
 'faze.',
 'Thisughhhh!',
 'kisage,',
 'noughhhh!',
 'ughhh?mereugh?',
 'kenaar',
 'ughhh?ofugh?',
 'ughhh?kanizy,',
 'ughhh?is',
 'a',
 'keszige',
 'of',
 'zhe',
 'kox',
 'ughhh!populi',
 'now',
 'kacanz,',
 'kanished.ugh?',
 'However,',
 'zhis',
 'kalorous',
 'kisizazion',
 'of',
 'a',
 'bygone',
 'kexazion',
 'szands',
 'kivified,',
 'ughhh!and',
 'hasugh.',
 'kowed',
 'zoughhhh!',
 'kanquish',
 'zhese',
 'kenalugh?',
 'andughhhh!',
 'ughhh!kirulenz',
 'kermin,',
 'kan',
 'guarding',
 'kiceugh.',
 'and',
 'kouchsafingugh?',
 'zhe',
 'kiolenzlb',
 'kicious',
 'ughhh?and',
 'koraciousugh?',
 'kiolazion',
 'of',
 'kolizion.ugh?',
 'ughhh?Theugh.',
 'onlb',
 'kerdicz',
 'isugh.',
 'kengeance;',
 'a',
 'ughhh?kendezza,',
 'heldughhhh!',
 'ughhh!as',
 'a',
 'kozive',
 'noz',
 'in',

2. Any time you see the word "ugh", with any number of h's, followed by a punctuation mark, delete this.

In [9]:
nuh_ugh = [re.sub(r'ugh+h*[!?.]*', '', i) for i in list_message]
nuh_ugh

['Koila!',
 'In',
 'kiew,',
 'a',
 'humble',
 'kaudevillianaoghajdbn',
 'kezeran,',
 'casz',
 'kicariouslb',
 'as',
 'bozh',
 'kiczim',
 'and',
 'killain',
 'bb',
 'zhe',
 'kicissizudes',
 'of',
 'faze.',
 'This',
 'kisage,',
 'no',
 'mere',
 'kenaar',
 'of',
 'kanizy,',
 'is',
 'a',
 'keszige',
 'of',
 'zhe',
 'kox',
 'populi',
 'now',
 'kacanz,',
 'kanished.',
 'However,',
 'zhis',
 'kalorous',
 'kisizazion',
 'of',
 'a',
 'bygone',
 'kexazion',
 'szands',
 'kivified,',
 'and',
 'has',
 'kowed',
 'zo',
 'kanquish',
 'zhese',
 'kenal',
 'and',
 'kirulenz',
 'kermin,',
 'kan',
 'guarding',
 'kice',
 'and',
 'kouchsafing',
 'zhe',
 'kiolenzlb',
 'kicious',
 'and',
 'koracious',
 'kiolazion',
 'of',
 'kolizion.',
 'The',
 'onlb',
 'kerdicz',
 'is',
 'kengeance;',
 'a',
 'kendezza,',
 'held',
 'as',
 'a',
 'kozive',
 'noz',
 'in',
 'kain,',
 'for',
 'zhe',
 'kalue',
 'and',
 'keracizb',
 'of',
 'such',
 'shall',
 'one',
 'dab',
 'kindicaze',
 'zhe',
 'kigilanz',
 'and',
 'zhe',
 'kirzuous

3. No word should be longer than 16 characters. Drop all extra characters beyond 13 off the end of each word.

In [10]:
nuh_ugh = [''.join(word[:13]) for word in nuh_ugh]
nuh_ugh

['Koila!',
 'In',
 'kiew,',
 'a',
 'humble',
 'kaudevilliana',
 'kezeran,',
 'casz',
 'kicariouslb',
 'as',
 'bozh',
 'kiczim',
 'and',
 'killain',
 'bb',
 'zhe',
 'kicissizudes',
 'of',
 'faze.',
 'This',
 'kisage,',
 'no',
 'mere',
 'kenaar',
 'of',
 'kanizy,',
 'is',
 'a',
 'keszige',
 'of',
 'zhe',
 'kox',
 'populi',
 'now',
 'kacanz,',
 'kanished.',
 'However,',
 'zhis',
 'kalorous',
 'kisizazion',
 'of',
 'a',
 'bygone',
 'kexazion',
 'szands',
 'kivified,',
 'and',
 'has',
 'kowed',
 'zo',
 'kanquish',
 'zhese',
 'kenal',
 'and',
 'kirulenz',
 'kermin,',
 'kan',
 'guarding',
 'kice',
 'and',
 'kouchsafing',
 'zhe',
 'kiolenzlb',
 'kicious',
 'and',
 'koracious',
 'kiolazion',
 'of',
 'kolizion.',
 'The',
 'onlb',
 'kerdicz',
 'is',
 'kengeance;',
 'a',
 'kendezza,',
 'held',
 'as',
 'a',
 'kozive',
 'noz',
 'in',
 'kain,',
 'for',
 'zhe',
 'kalue',
 'and',
 'keracizb',
 'of',
 'such',
 'shall',
 'one',
 'dab',
 'kindicaze',
 'zhe',
 'kigilanz',
 'and',
 'zhe',
 'kirzuous.',
 'Ke

4. Replace all instances of exactly 2 a's with exactly 2 e's.

In [11]:
def replace_aa(word):
    return word.replace("aa", "ee")

nuh_ugh = [replace_aa(word) for word in nuh_ugh]
nuh_ugh

['Koila!',
 'In',
 'kiew,',
 'a',
 'humble',
 'kaudevilliana',
 'kezeran,',
 'casz',
 'kicariouslb',
 'as',
 'bozh',
 'kiczim',
 'and',
 'killain',
 'bb',
 'zhe',
 'kicissizudes',
 'of',
 'faze.',
 'This',
 'kisage,',
 'no',
 'mere',
 'keneer',
 'of',
 'kanizy,',
 'is',
 'a',
 'keszige',
 'of',
 'zhe',
 'kox',
 'populi',
 'now',
 'kacanz,',
 'kanished.',
 'However,',
 'zhis',
 'kalorous',
 'kisizazion',
 'of',
 'a',
 'bygone',
 'kexazion',
 'szands',
 'kivified,',
 'and',
 'has',
 'kowed',
 'zo',
 'kanquish',
 'zhese',
 'kenal',
 'and',
 'kirulenz',
 'kermin,',
 'kan',
 'guarding',
 'kice',
 'and',
 'kouchsafing',
 'zhe',
 'kiolenzlb',
 'kicious',
 'and',
 'koracious',
 'kiolazion',
 'of',
 'kolizion.',
 'The',
 'onlb',
 'kerdicz',
 'is',
 'kengeance;',
 'a',
 'kendezza,',
 'held',
 'as',
 'a',
 'kozive',
 'noz',
 'in',
 'kain,',
 'for',
 'zhe',
 'kalue',
 'and',
 'keracizb',
 'of',
 'such',
 'shall',
 'one',
 'dab',
 'kindicaze',
 'zhe',
 'kigilanz',
 'and',
 'zhe',
 'kirzuous.',
 'Ke

5. Replace all z's with t's.

In [12]:
def replace_z(word):
    return word.replace("z", "t")

nuh_ugh = [replace_z(word) for word in nuh_ugh]
nuh_ugh

['Koila!',
 'In',
 'kiew,',
 'a',
 'humble',
 'kaudevilliana',
 'keteran,',
 'cast',
 'kicariouslb',
 'as',
 'both',
 'kictim',
 'and',
 'killain',
 'bb',
 'the',
 'kicissitudes',
 'of',
 'fate.',
 'This',
 'kisage,',
 'no',
 'mere',
 'keneer',
 'of',
 'kanity,',
 'is',
 'a',
 'kestige',
 'of',
 'the',
 'kox',
 'populi',
 'now',
 'kacant,',
 'kanished.',
 'However,',
 'this',
 'kalorous',
 'kisitation',
 'of',
 'a',
 'bygone',
 'kexation',
 'stands',
 'kivified,',
 'and',
 'has',
 'kowed',
 'to',
 'kanquish',
 'these',
 'kenal',
 'and',
 'kirulent',
 'kermin,',
 'kan',
 'guarding',
 'kice',
 'and',
 'kouchsafing',
 'the',
 'kiolentlb',
 'kicious',
 'and',
 'koracious',
 'kiolation',
 'of',
 'kolition.',
 'The',
 'onlb',
 'kerdict',
 'is',
 'kengeance;',
 'a',
 'kendetta,',
 'held',
 'as',
 'a',
 'kotive',
 'not',
 'in',
 'kain,',
 'for',
 'the',
 'kalue',
 'and',
 'keracitb',
 'of',
 'such',
 'shall',
 'one',
 'dab',
 'kindicate',
 'the',
 'kigilant',
 'and',
 'the',
 'kirtuous.',
 'Ke

6. Every word that ends in b, change that to a y. Hint: look out for punctuation!

In [ ]:
nuh_ugh = list(map(lambda x: re.sub(r'b\b', 'y', x), nuh_ugh))
nuh_ugh

7. Every word that starts with k, change that to a v. Hint: look out for capitalization!

In [18]:
nuh_ugh = list(map(lambda x: re.sub(r'^k', 'v', x), nuh_ugh))
nuh_ugh = list(map(lambda x: re.sub(r'^K', 'v', x), nuh_ugh))
nuh_ugh

['voila!',
 'In',
 'view,',
 'a',
 'humble',
 'vaudevilliana',
 'veteran,',
 'cast',
 'vicariously',
 'as',
 'both',
 'victim',
 'and',
 'villain',
 'by',
 'the',
 'vicissitudes',
 'of',
 'fate.',
 'This',
 'visage,',
 'no',
 'mere',
 'veneer',
 'of',
 'vanity,',
 'is',
 'a',
 'vestige',
 'of',
 'the',
 'vox',
 'populi',
 'now',
 'vacant,',
 'vanished.',
 'However,',
 'this',
 'valorous',
 'visitation',
 'of',
 'a',
 'bygone',
 'vexation',
 'stands',
 'vivified,',
 'and',
 'has',
 'vowed',
 'to',
 'vanquish',
 'these',
 'venal',
 'and',
 'virulent',
 'vermin,',
 'van',
 'guarding',
 'vice',
 'and',
 'vouchsafing',
 'the',
 'violently',
 'vicious',
 'and',
 'voracious',
 'violation',
 'of',
 'volition.',
 'The',
 'only',
 'verdict',
 'is',
 'vengeance;',
 'a',
 'vendetta,',
 'held',
 'as',
 'a',
 'votive',
 'not',
 'in',
 'vain,',
 'for',
 'the',
 'value',
 'and',
 'veracity',
 'of',
 'such',
 'shall',
 'one',
 'day',
 'vindicate',
 'the',
 'vigilant',
 'and',
 'the',
 'virtuous.',
 've

8. Use .join() to recombine all your words into a message.

In [19]:
' '.join(nuh_ugh)

'voila! In view, a humble vaudevilliana veteran, cast vicariously as both victim and villain by the vicissitudes of fate. This visage, no mere veneer of vanity, is a vestige of the vox populi now vacant, vanished. However, this valorous visitation of a bygone vexation stands vivified, and has vowed to vanquish these venal and virulent vermin, van guarding vice and vouchsafing the violently vicious and voracious violation of volition. The only verdict is vengeance; a vendetta, held as a votive not in vain, for the value and veracity of such shall one day vindicate the vigilant and the virtuous. verily this vichyssoise of verbiage veers most verbose, so let me simply add that its my very good honour to meet you and you may call me v.'